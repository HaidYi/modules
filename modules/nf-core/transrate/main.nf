process TRANSRATE {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/5c/5c7d34ed4d4e8b30bc4216c753fcf3084d3bab43f8f2e80eb63b6184fbd25f04/data':
        'community.wave.seqera.io/library/snap-aligner_transrate:501869af4f81472b' }"

    input:
    tuple val(meta) , path(fasta)     // assembly file
    tuple val(meta2), path(reads)     // processed reads
    tuple val(meta3), path(reference) // reference proteins or transcripts fasta

    output:
    tuple val(meta), path("${prefix}"), emit: transrate_results
    path "versions.yml"               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    def transrate_assembly = fasta ? "--assembly $fasta" : ''
    def transrate_reads = reads ? ( meta2.single_end ? "--left $reads" : "--left ${reads[0]} --right ${reads[1]}" ) : "" // TODO test with meta.single_end
    def transrate_reference = reference ? "--reference $reference" : ''
    """
    transrate \\
        $args \\
        $transrate_assembly \\
        $transrate_reads \\
        $transrate_reference \\
        --output ${prefix}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        transrate: \$(transrate --version 2>/dev/null | tail -n1)
        snapaligner: \$(snap-aligner 2>&1| head -n 1 | sed 's/^.*version //;s/.\$//')
        salmon: \$(salmon --version | sed -e "s/salmon //g")
        blast: \$(blastp -version 2>&1 | sed 's/^.*blastp: //; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    mkdir ${prefix}

    touch ${prefix}/test.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        transrate: \$(transrate --version 2>/dev/null | tail -n1)
        snapaligner: \$(snap-aligner 2>&1| head -n 1 | sed 's/^.*version //;s/.\$//')
        salmon: \$(salmon --version | sed -e "s/salmon //g")
        blast: \$(blastp -version 2>&1 | sed 's/^.*blastp: //; s/ .*\$//')
    END_VERSIONS
    """
}
