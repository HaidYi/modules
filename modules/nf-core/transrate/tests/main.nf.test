nextflow_process {

    name "Test Process TRANSRATE"
    script "../main.nf"
    process "TRANSRATE"

    tag "modules"
    tag "modules_nfcore"
    tag "transrate"
    tag "gunzip"

    test("sarscov2 - assembly - reference") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test_assembly'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = [ [ id:'test_reads', single_end:false ], [] ]
                input[2] = [
                    [ id:'test_reference' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/proteome.fasta', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    // test("sarscov2 - assembly - reads - reference") { // reads produce an error: Failed to build Snap index

    //     setup {
    //         run("GUNZIP", alias: "GUNZIP_LEFT") {
    //             script "../../gunzip/main.nf"
    //             process {
    //                 """
    //                 input[0] = [
    //                     [ id: 'test_reads', single_end:false ],
    //                     file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
    //                 ]
    //                 """
    //             }
    //         }

    //         run("GUNZIP", alias: "GUNZIP_RIGHT") {
    //             script "../../gunzip/main.nf"
    //             process {
    //                 """
    //                 input[0] = [
    //                     [ id: 'test_reads', single_end:false ],
    //                     file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
    //                 ]
    //                 """
    //             }
    //         }
    //     }

    //     when {
    //         process {
    //             """
    //             input[0] = [
    //                 [ id:'test_assembly'],
    //                 file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
    //             ]
    //             input[1] = GUNZIP_LEFT.out.gunzip.join(GUNZIP_RIGHT.out.gunzip).map{ meta, file1, file2 -> [ meta, [file1, file2] ] }
    //             input[2] = [
    //                 [ id:'test_reference' ],
    //                 file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/proteome.fasta', checkIfExists: true)
    //             ]
    //             """
    //         }
    //     }

    //     then {
    //         assertAll(
    //             { assert process.success },
    //             { assert snapshot(process.out).match() }
    //         )
    //     }

    // }

    test("sarscov2 - assembly - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test_assembly'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                input[1] = [ [ id:'test_reads', single_end:false ], [] ]
                input[2] = [ [ id:'test_reference' ], [] ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
