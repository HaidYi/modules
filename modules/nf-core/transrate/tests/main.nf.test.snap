{
    "sarscov2 - assembly - reference": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_assembly"
                        },
                        [
                            "assemblies.csv:md5,e888c259f97d17f6e6346403f463b9cb",
                            [
                                "contigs.csv:md5,675af88b31f4819be45a40ec5fc29f01",
                                "evalues_data:md5,d4db173f537be04127dbd7bab7eb344e",
                                "fitting_data:md5,efbc5b5ddae1f7afa09c95068bd48870",
                                "genome.nhr:md5,f4b4ddb034fd3dd7b25c89e9d50c004e",
                                "genome.nin:md5,9638fb07bf668722dbfe546f7b291577",
                                "genome.nsq:md5,982cbc7d9e38743b9b1037588862b9da",
                                "genome_into_proteome.1.blast:md5,69f94299b3098e7685e016e454414662",
                                "proteome.phr:md5,c523f92ddfe648ae41ef2a17c599937f",
                                "proteome.pin:md5,fb0864d566b95d5d0f9145e8c70f610e",
                                "proteome.psq:md5,026b799b20282170a0b350c7aa7b9fe0",
                                "proteome_into_genome.2.blast:md5,c457796316a035f9dae60bf160269126"
                            ]
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,e4f01fdcbc4b3acf0c749bf8749523bd"
                ],
                "transrate_results": [
                    [
                        {
                            "id": "test_assembly"
                        },
                        [
                            "assemblies.csv:md5,e888c259f97d17f6e6346403f463b9cb",
                            [
                                "contigs.csv:md5,675af88b31f4819be45a40ec5fc29f01",
                                "evalues_data:md5,d4db173f537be04127dbd7bab7eb344e",
                                "fitting_data:md5,efbc5b5ddae1f7afa09c95068bd48870",
                                "genome.nhr:md5,f4b4ddb034fd3dd7b25c89e9d50c004e",
                                "genome.nin:md5,9638fb07bf668722dbfe546f7b291577",
                                "genome.nsq:md5,982cbc7d9e38743b9b1037588862b9da",
                                "genome_into_proteome.1.blast:md5,69f94299b3098e7685e016e454414662",
                                "proteome.phr:md5,c523f92ddfe648ae41ef2a17c599937f",
                                "proteome.pin:md5,fb0864d566b95d5d0f9145e8c70f610e",
                                "proteome.psq:md5,026b799b20282170a0b350c7aa7b9fe0",
                                "proteome_into_genome.2.blast:md5,c457796316a035f9dae60bf160269126"
                            ]
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e4f01fdcbc4b3acf0c749bf8749523bd"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-05-07T18:35:18.517991541"
    },
    "sarscov2 - assembly - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_assembly"
                        },
                        [
                            "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,e4f01fdcbc4b3acf0c749bf8749523bd"
                ],
                "transrate_results": [
                    [
                        {
                            "id": "test_assembly"
                        },
                        [
                            "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e4f01fdcbc4b3acf0c749bf8749523bd"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-05-07T18:35:27.022306799"
    }
}